User manual: docs for copy/paste blocks in Volto

MAX-786 opened the pull request:
#3827 

docs for copy-pasting blocks in Volto.


https://user-images.githubusercontent.com/99530996/213858917-72afebeb-b4d2-4952-8d87-5267978d1560.mp4

MAX-786 commented:
@stevepiercy, Does this also come under the User manual? 

Also, IMHO we should move all docs related to `blocks` one more level deep like : (since other content needs to be at outer level)
```{js}

|-- User Manual
    |-- Volto Blocks
        |-- Edit content using Blocks
        |-- Copy, Cut, and Paste Blocks
        |-- Drag Blocks
        |--  ... ... ...
    |-- Move Content in Volto
    |-- Undo Control Panel
    | -- ... ... ...
```

tiberiuichim commented:
Awesome! Some suggestions:

- the screenshots may look a bit weird being so elongated. Maybe you could include some blocks on the right side (the default Volto intro page is fine) and visually highlight the block clipboard buttons.
- there's additional capabilities with the copy/pasting of blocks. After you select some blocks (so, hit ctrl key, click start / end block), you can additionally hit the shift key and you'll be able to selectively deselect some of the selected blocks. In adition, if you hold the ctrl key while clicking the paste button, it doesn't clear the clipboard.

stevepiercy commented:
> @stevepiercy, Does this also come under the User manual?

Yes, indeed!

> Also, IMHO we should move all docs related to `blocks` one more level deep

"[Flat is better than nested.](https://peps.python.org/pep-0020/#the-zen-of-python)" Let's keep pages at the root of the User Manual until it becomes obvious that deeper levels are needed.

One of the problems of Plone 5 docs is its deeply-nested organization, which makes it very difficult to find anything. We do not want to repeat that.

MAX-786 commented:
> the screenshots may look a bit weird being so elongated. Maybe you could include some blocks on the right side (the default Volto intro page is fine) and visually highlight the block clipboard buttons.

Sure! means we should use a whole page screenshot rather than just elongated Toolbar Screenshot.

>  In adition, if you hold the ctrl key while clicking the paste button, it doesn't clear the clipboard.

Oh man!!! only if I knew this feature before that would've made my life sooo easier...... gotta document these hidden features asap. 🙂

MAX-786 commented:
> One of the problems of Plone 5 docs is its deeply-nested organization, which makes it very difficult to find anything. We do not want to repeat that.

Oh I understand, lets leave it flat for now and I'll doc suggested features and notify you when PR is ready for review.

MAX-786 commented:
Thanks😊 @tiberiuichim, for your suggestion and mentioning that hidden feature, I guess it looks good now and covers mostly all things that this PR was intended for. Feedback Welcome.

sneridagh commented:
@stevepiercy any followup on this one? please give me your ack on this one.

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
````diff
@@ -0,0 +1,43 @@
+This feature can be used by selecting a start block and an end block while holding the <kbd>shift</kbd> key.
+This will select all the blocks between the start and end block, allowing you to copy or delete multiple blocks at once.
+
+```{image} ../_static/user-manual/blocks/block-copy-cut.gif
````
Animations are poor for usability, accessibility, and file size. We should instead use videos that have full playback capability and may be loaded async. I admit I let a few sneak in before, but I now realize we should resolve to improve the usability.

Also the recordings get reduced to a narrow viewport width of 686 pixels on desktop, and less on mobile. When recording we should try to keep the browser width, or the recorded area, to 680 pixels maximum.

mp4 is a well supported video format.

Here's the Sphinx extension that needs to be installed.

https://github.com/sphinx-contrib/video

MAX-786 replied:
Thanks for the input! I thought the same, so I'll re-record it and upload the video with MP4 extension installed.

MAX-786 replied:
@stevepiercy, Do you have any suggestions on how I can improve the animations in the video?

stevepiercy replied:
Before I start recording, I usually resize the browser window to as narrow as possible to reduce unused space and focus attention on the action.

I took two screenshots. Note that I have clicked the vertical handle to collapse the right panel to make more space.

Note that the copy/cut toolbar is at the top of the browser window for 680 pixels, and drops to the side at 768 pixels. We need to be mindful of different device viewport sizes.

## 680 pixels wide

![Screen Shot 2023-02-03 at 5 33 57 AM](https://user-images.githubusercontent.com/102112/216616227-4884a49f-5d02-45af-bc36-922b51931352.png)

## 768 pixels wide

![Screen Shot 2023-02-03 at 5 38 04 AM](https://user-images.githubusercontent.com/102112/216617039-786adc28-cc16-40e9-9c5f-44fd15185b76.png)

I think that setting the browser width to 768 pixels will accommodate the largest audience of users. I doubt most users will edit with a phone or tablet, but we can always revisit that audience in future updates. I would focus on the desktop audience for now.

We also might update our documentation theme to accommodate wider layouts for the main content area, but that is another future consideration, not something to consider now.

MAX-786 replied:
Thanks for the update! , I'll implement the suggested changes soon.
> We also might update our documentation theme to accommodate wider layouts for the main content area, but that is another future consideration, not something to consider now.

Ohh.. sure love to look into it in near future :)

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+---
+myst:
+  html_meta:
+    "description": "Copy, cut and paste blocks in Volto, the Plone 6 frontend."
```
```suggestion
    "description": "Copy, cut, and paste blocks in Volto, the Plone 6 frontend."
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+myst:
+  html_meta:
+    "description": "Copy, cut and paste blocks in Volto, the Plone 6 frontend."
+    "property=og:description": "Copy, cut and paste blocks in Volto, the Plone 6 frontend."
```
```suggestion
    "property=og:description": "Copy, cut, and paste blocks in Volto, the Plone 6 frontend."
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+
+# Copy, Cut, and Paste blocks in Volto
+
+Volto supports copying/cutting of blocks from one page and pasting them to another page. Here's how you can do:
```
```suggestion
Volto supports copying and cutting of blocks from one page, and pasting them to another page.
This chapter describes how to do so.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+
+## Copy/Cut blocks
+
+While in edit mode you can hold <kbd>ctrl</kbd> key to select multiple blocks you want to copy/cut.
```
Avoid inserting HTML into Markdown because it is not supported in PDF and other output formats. Use MyST markup instead.

```suggestion
While in edit mode you can hold :kbd:`ctrl` key to select multiple blocks you want to copy/cut.
```

MAX-786 replied:
Oh... Thanks Steve! didn't knew that. Noted for next time :)

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+## Copy/Cut blocks
+
+While in edit mode you can hold <kbd>ctrl</kbd> key to select multiple blocks you want to copy/cut.
+Once blocks are selected, you will have a copy <i class="fas fa-copy"></i> and a cut <i class="fas fa-cut"></i> option at the bottom of Toolbar.
```
We should use the [Volto (Pastanaga) icons](https://pastanaga.io/icons/), not Font Awesome icons. We want any icons used inline to be consistent with the actual UI.

This also requires a new MyST extension. See https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#html-images

Here is untested sample markup. Also I know I just said that HTML should be avoided, but that MyST extension supposedly resolves the issue of different output formats.

```suggestion
Once blocks are selected, you will have a copy <img alt="../_static/copy.svg" src="../_static/copy.svg" style="height: 1em;"> and a cut <img alt="../_static/cut.svg" src="../_static/cut.svg" style="height: 1em;"> option at the bottom of Toolbar.
```

MAX-786 replied:
Sure will do!

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+Once blocks are selected, you will have a copy <i class="fas fa-copy"></i> and a cut <i class="fas fa-cut"></i> option at the bottom of Toolbar.
+
+Additionally, Volto also has a hidden feature for selecting multiple blocks in sequence.
+This feature can be used by selecting a start block and an end block while holding the <kbd>shift</kbd> key.
```
```suggestion
This feature can be used by selecting a start block and an end block while holding the `kbd`shift` key.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+
+Additionally, Volto also has a hidden feature for selecting multiple blocks in sequence.
+This feature can be used by selecting a start block and an end block while holding the <kbd>shift</kbd> key.
+This will select all the blocks between the start and end block, allowing you to copy or delete multiple blocks at once.
```
```suggestion
This will select all the blocks between the start and end blocks, allowing you to copy or delete multiple blocks at once.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+
+## Paste blocks
+
+Once you copy/cut the blocks you can paste those blocks on that page or some other page.
```
```suggestion
Once you copy or cut the blocks, you can paste those blocks on that page or some other page.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+## Paste blocks
+
+Once you copy/cut the blocks you can paste those blocks on that page or some other page.
+Select an empty block where you want to paste blocks from clipboard.
```
```suggestion
Select an empty block where you want to paste blocks from the clipboard.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+
+Once you copy/cut the blocks you can paste those blocks on that page or some other page.
+Select an empty block where you want to paste blocks from clipboard.
+You can see the paste option <i class="fas fa-clipboard"></i> at the bottom of Toolbar.
```
```suggestion
You can see the paste option <img alt="../_static/paste.svg" src="../_static/paste.svg" style="height: 1em;"> at the bottom of the toolbar.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,43 @@
+Select an empty block where you want to paste blocks from clipboard.
+You can see the paste option <i class="fas fa-clipboard"></i> at the bottom of Toolbar.
+
+Also, in addition if you hold <kbd>ctrl</kbd> key while clicking the paste button, it does not clear the clipboard.
```
```suggestion
Also if you hold the :kbd:`ctrl` key while clicking the paste button, it does not clear the clipboard.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
````diff
@@ -0,0 +1,43 @@
+
+Also, in addition if you hold <kbd>ctrl</kbd> key while clicking the paste button, it does not clear the clipboard.
+
+```{image} ../_static/user-manual/blocks/block-paste.gif
````
Let's replace the gif with a video.

stevepiercy commented on news/3827.documentation:
```diff
@@ -0,0 +1 @@
+Add documentation for copy/paste blocks in Volto. @MAX-786
```
```suggestion
Add documentation for copy, cut, and paste blocks in Volto. @MAX-786
```

stevepiercy reviewed (changes requested):
This is impressive. Thank you!

You've also brought up two new things in documentation.

1. We need to standardize videos for accessibility.
2. We need to allow inline images.

Both of these new things need to be added to the `plone/documentation` repo and documented for authors, but we can do that later and separately.

I also made some suggestions for English grammar.

MAX-786 commented:
@stevepiercy or @ksuess, I understand that you both might be busy with WORK but whenever its fine can anyone of you please help me properly install `sphinxcontrib.video` extension😊. What I did :
1. Added `sphinxcontrib.video` to `conf.py`
2. Added it to `requirements-docs.txt` in volto repo.

Even after doing this extension is not working. Do I need to add it to the documentation repo's `requirement.txt` too? Thanks😊!

stevepiercy commented on requirements-docs.txt:
```diff
@@ -11,3 +11,4 @@ sphinx-sitemap
 sphinx-togglebutton
 sphinxcontrib-spelling
 sphinxext-opengraph
+sphinxcontrib.video
```
The correct package name is `sphinxcontrib-video`, whereas the extension name is `sphinxcontrib.video`. I know...

Then, of course, you can either do a `pip install sphinxcontrib-video` or any `make` command to install it.

```suggestion
sphinxcontrib-video
```

stevepiercy reviewed (changes requested):
Use correct package name.

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,40 @@
+## Copy/Cut blocks
+
+While in edit mode you can hold `ctrl` key to select multiple blocks you want to copy/cut.
+Once blocks are selected, you will have a copy <img alt="../_static/copy.svg" src="../_static/copy.svg" style="height: 1em;"> and a cut <img alt="../_static/cut.svg" src="../_static/cut.svg" style="height: 1em;"> option at the bottom of Toolbar.
```
sphinx-book-theme has an additional style that we want to override for inline images.

```css
img {
    margin:1rem 0;
}
```

So let's add a custom class.

```css
/* override sphinx-book-theme img tag with a class */
img.inline {
    margin: 0;
    height: 1em;
}
```

If that does not work, then add ` !important` to the styles as needed.

Also modify the inline HTML.

```suggestion
Once blocks are selected, you will have a copy <img alt="../_static/copy.svg" src="../_static/copy.svg" class="inline"> and a cut <img alt="../_static/cut.svg" src="../_static/cut.svg" class="inline"> option at the bottom of Toolbar.
```

stevepiercy replied:
I should add, I don't know if that MyST extension is doing what it promises. Let's make sure it does in fact add ` class="inline"` to the `<img>` tag when it is rendered.

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,40 @@
+
+Once you copy or cut the blocks, you can paste those blocks on that page or some other page.
+Select an empty block where you want to paste blocks from the clipboard.
+You can see the paste option <img alt="../_static/paste.svg" src="../_static/paste.svg" style="height: 1em;"> at the bottom of the toolbar.
```
```suggestion
You can see the paste option <img alt="../_static/paste.svg" src="../_static/paste.svg" class="inline"> at the bottom of the toolbar.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,40 @@
+This feature can be used by selecting a start block and an end block while holding the `shift` key.
+This will select all the blocks between the start and end blocks, allowing you to copy or delete multiple blocks at once.
+
+.. video:: ../_static/user-manual/blocks/block-copy-cut.mp4
```
This needs to be in MyST syntax

````suggestion
```{eval-rst}
.. video:: ../_static/user-manual/blocks/block-copy-cut.mp4
```
````

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,40 @@
+
+Also if you hold the `ctrl` key while clicking the paste button, it does not clear the clipboard.
+
+.. video:: ../_static/user-manual/blocks/block-paste.mp4
```
````suggestion
```{eval-rst}
.. video:: ../_static/user-manual/blocks/block-paste.mp4
```
````

stevepiercy reviewed (changes requested):
One MyST syntax change, and try using a class instead of a style for the Pastanaga icons.

stevepiercy commented:
BTW, nice width and reduction in file size of the mp4 videos to 10% of the original gifs, the latter of which may be deleted from the PR. Users will appreciate this! They are slightly pixelated though, so could you make the compression less aggressive for a clearer image?

MAX-786 commented:
> BTW, nice width and reduction in file size of the mp4 videos to 10% of the original gifs, the latter of which may be deleted from the PR

yeah, I intentionally left original gifs there for taking reference. I'll remove it now.

>Users will appreciate this! They are slightly pixelated though, so could you make the compression less aggressive for a clearer image?

Sure! will do

MAX-786 commented:
Done!

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,46 @@
+
+(copy-cut-blocks-label)=
+
+## Copy/Cut blocks
```
```suggestion
## Copy or cut blocks
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,46 @@
+
+## Copy/Cut blocks
+
+While in edit mode you can hold {kbd}`ctrl` key to select multiple blocks you want to copy/cut.
```
```suggestion
While in edit mode, you can hold the {kbd}`ctrl` key to select multiple blocks that you want to copy or cut.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,46 @@
+## Copy/Cut blocks
+
+While in edit mode you can hold {kbd}`ctrl` key to select multiple blocks you want to copy/cut.
+Once blocks are selected, you will have a copy <img alt="../_static/copy.svg" src="../_static/copy.svg" class="inline"> and a cut <img alt="../_static/cut.svg" src="../_static/cut.svg" class="inline"> option at the bottom of Toolbar.
```
The toolbar goes horizontal and to the top of the window on small screens.

```suggestion
Once blocks are selected, you will have copy <img alt="../_static/copy.svg" src="../_static/copy.svg" class="inline"> and cut <img alt="../_static/cut.svg" src="../_static/cut.svg" class="inline"> options in the toolbar.
```

MAX-786 replied:
Oh yeah I missed that , Thanks for pointing it out.

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,46 @@
+While in edit mode you can hold {kbd}`ctrl` key to select multiple blocks you want to copy/cut.
+Once blocks are selected, you will have a copy <img alt="../_static/copy.svg" src="../_static/copy.svg" class="inline"> and a cut <img alt="../_static/cut.svg" src="../_static/cut.svg" class="inline"> option at the bottom of Toolbar.
+
+Additionally, Volto also has a hidden feature for selecting multiple blocks in sequence.
```
```suggestion
Volto also allows the selection of multiple contiguous blocks.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,46 @@
+
+Once you copy or cut the blocks, you can paste those blocks on that page or some other page.
+Select an empty block where you want to paste blocks from the clipboard.
+You can see the paste option <img alt="../_static/paste.svg" src="../_static/paste.svg" class="inline"> at the bottom of the toolbar.
```
```suggestion
You can see the paste option <img alt="../_static/paste.svg" src="../_static/paste.svg" class="inline"> in the toolbar.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,46 @@
+Select an empty block where you want to paste blocks from the clipboard.
+You can see the paste option <img alt="../_static/paste.svg" src="../_static/paste.svg" class="inline"> at the bottom of the toolbar.
+
+Also if you hold the {kbd}`ctrl` key while clicking the paste button, it does not clear the clipboard.
```
```suggestion
Also if you hold the {kbd}`ctrl` key while clicking the paste button, it keeps the clipboard buffer, allowing you to repeatedly paste it.
```

stevepiercy reviewed (changes requested):
This is awesome!

I made a few grammar suggestions, and once those are corrected, this is good to go. Thank you!

MAX-786 commented:
Thanks a lot, @stevepiercy for your time, and also thanks @tiberiuichim for suggesting hidden features 😊.

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,46 @@
+
+Volto also allows the selection of multiple contiguous blocks.
+This feature can be used by selecting a start block and an end block while holding the {kbd}`shift` key.
+This will select all the blocks between the start and end blocks, allowing you to copy or delete multiple blocks at once.
```
```suggestion
This will select all the blocks between the start and end blocks, allowing you to copy, cut, or delete multiple blocks at once.
```

stevepiercy commented on docs/source/user-manual/copy-paste-blocks.md:
```diff
@@ -0,0 +1,46 @@
+
+Once you copy or cut the blocks, you can paste those blocks on that page or some other page.
+Select an empty block where you want to paste blocks from the clipboard.
+You can see the paste option <img alt="../_static/paste.svg" src="../_static/paste.svg" class="inline"> in the toolbar.
```
```suggestion
You can click the paste option <img alt="../_static/paste.svg" src="../_static/paste.svg" class="inline"> in the toolbar.
```

stevepiercy reviewed (changes requested):
Two quick changes, which I will commit now, then merge. Thank you!

stevepiercy reviewed (approved):
This is ready to merge as soon as it passes the final run.

stevepiercy commented:
@MAX-786 the next step is to update the configuration in `plone/documentation`. I cannot deploy this PR to production until the main repo is configured to handle videos and inline images. Would you please be so kind to start work on that?

MAX-786 commented:
> @MAX-786 the next step is to update the configuration in `plone/documentation`. I cannot deploy this PR to production until the main repo is configured to handle videos and inline images. Would you please be so kind to start work on that?

Sure, I'll look into it tomorrow. Thanks for suggesting it to me😄.

Just confirming what i basically need to do in `documentation` repo:
- add `html-image` and `sphinxcontrib.video` extension in "conf.py" of that repo.
- add `sphinxcontrib-video` in requirements.txt.

IG this will be enough to configure it. Please just let me know if i am missing something. Thanks!

stevepiercy commented:
You got it! I created the new issue https://github.com/plone/documentation/issues/1436
